### DISCUSSION TITLE
Where are the l10n strings in the output object in `v3`?

### DISCUSSION THREAD

**Discussion by tukusejssirs** [romcal/romcal#303] [Q&A] [answered]
I cannot find the localised strings in the romcal object. I’ve thought they are in `i18nDef`, but there is only the key name.

```js
// Code
const Romcal = require('romcal')
const {Slovakia_Sk} = require('../bundles/slovakia/cjs/sk')

async function main() {
  const romcal = new Romcal({localizedCalendar: Slovakia_Sk})
  const r = await romcal.generateCalendar(2021)
  console.log(r['2021-01-01'])
}

// Output
[
  Za {
    key: 'mary_mother_of_god',
    date: '2021-01-01',
    dateDef: { dateFn: 'maryMotherOfGod', yearOffset: 0 },
    precedence: 'GENERAL_SOLEMNITY_3',
    rank: 'SOLEMNITY',
    isHolyDayOfObligation: true,
    isOptional: false,
    i18nDef: [ 'names:mary_mother_of_god' ],
    seasons: [ 'CHRISTMAS_TIME' ],
    periods: [ 'CHRISTMAS_OCTAVE', 'CHRISTMAS_TO_PRESENTATION_OF_THE_LORD' ],
    colors: [ 'WHITE' ],
    martyrology: [],
    titles: [],
    calendar: {
      weekOfSeason: 2,
      dayOfSeason: 8,
      dayOfWeek: 5,
      nthDayOfWeekInMonth: 1,
      startOfSeason: '2020-12-25',
      endOfSeason: '2021-01-10',
      startOfLiturgicalYear: '2020-11-29',
      endOfLiturgicalYear: '2021-11-27'
    },
    cycles: {
      properCycle: 'PROPER_OF_TIME',
      sundayCycle: 'YEAR_B',
      weekdayCycle: 'YEAR_1',
      psalterWeek: 'WEEK_2'
    },
    fromCalendar: 'ProperOfTime',
    fromExtendedCalendars: []
  }
]
```

Do I need to `require` an additional file (compiled `sk.ts`)?

**Comment by emagnier** [comment]
Some properties like `.name`, `.seasonNames`, `.rankName`, `.colorNames` are not output by default in the `LiturgicalDay` object, because they are "lazy-computed".

So they are still accessible, but their values are computed only when you access these properties. In fact, everything related to l10n / i18n are costly in terms of performance, this is the reason why they are not computed by default.

```js
console.log(r['2021-01-01'][0].colorNames) // output [ "biela" ]
```

**Comment by emagnier** [comment]
However, you are flagging an important point. If you are using romcal as a middleware in a backend Rest API, these properties will also not be output by default. In this situation, an optional romcal option should be added, to specify that romcal must compute everything by default (no lazy computing). This will have performance impact, but it's easier to work around this issue in the backend world by using API cache, for example.

**Comment by tukusejssirs** [comment]
> Some properties like `.name`, `.seasonNames`, `.rankName`, `.colorNames` are not output by default in the `LiturgicalDay` object, because they are "lazy-computed".
> 
> So they are still accessible, but their values are computed only when you access these properties. In fact, everything related to l10n / i18n are costly in terms of performance, this is the reason why they are not computed by default.
> 
> ```js
> console.log(r['2021-01-01'][0].colorNames) // output [ "biela" ]
> ```

---

**Update:** My bad, missed `[0]` in everywhere below. :man_facepalming:

---

Well, I understand how it _should_ work, but it does not work in my case. For the mentioned properties, the output is `undefined`.

```js
const Romcal = require('romcal')
const {Slovakia_Sk} = require('../bundles/slovakia/cjs/sk')

main()

async function main() {
  const romcal = new Romcal({localizedCalendar: Slovakia_Sk})
  const r = await romcal.generateCalendar(2021)
  console.log(r['2021-01-01'].name)         // Output: undefined
  console.log(r['2021-01-01'].seasonNames)  // Output: undefined
  console.log(r['2021-01-01'].rankName)     // Output: undefined
  console.log(r['2021-01-01'].colorNames)   // Output: undefined
}
```

**Comment by tukusejssirs** [comment]
> However, you are flagging an important point. If you are using romcal as a middleware in a backend Rest API, these properties will also not be output by default. In this situation, an optional romcal option should be added, to specify that romcal must compute everything by default (no lazy computing). This will have performance impact, but it's easier to work around this issue in the backend world by using API cache, for example.

We might want to rename this issue and use it to add this option to romcal.

**Comment by emagnier** [comment, marked as answer]
I'm not sure if you finally made it working or not. Just in case, here are the steps to make it run.

And FYI, I've manually published a new NPM version of romcal, as well as all calendar plugins. I encourage you to install & require them instead.

```bash
# Create a new project
mkdir my-project
cd my-project
npm init -y

# Install romcal, and the Slovak calendar
npm install -d romcal @romcal/calendar.slovakia

# Create a JS file to run romcal
touch index.js
```

Then, your `index.js` should look like:
```js
const Romcal = require('romcal')
const {Slovakia_Sk} = require('@romcal/calendar.slovakia')

async function main() {
  const romcal = new Romcal({localizedCalendar: Slovakia_Sk})
  const r = await romcal.generateCalendar(2021)
  console.log(r['2021-01-01'][0].name)         // Output: Panny Márie Bohorodičky
  console.log(r['2021-01-01'][0].seasonNames)  // Output: [ 'Vianočné obdobie' ]
  console.log(r['2021-01-01'][0].rankName)     // Output: slávnosť
  console.log(r['2021-01-01'][0].colorNames)   // Output: [ 'biela' ]
}

main()
```

To run it:
```bash
node index.js
```

The `[0]` is indeed important, it's the first item in the array, that corresponds to the default `LiturgicalDay` object for the given date. All other following `LiturgicalDay` objects are most of the time optional memorials.

**Comment by emagnier** [comment]
> We might want to rename this issue and use it to add this option to romcal.

No, because it's 2 different topics, and it's good to don't mix them. I've created another issue for that: #300.

**Comment by tukusejssirs** [comment]
> I'm not sure if you finally made it working or not. Just in case, here are the steps to make it run.

Yes, I made it with your help. :wink: Thanks!

> And FYI, I've manually published a new NPM version of romcal, as well as all calendar plugins. I encourage you to install & require them instead.

Well, yeah, it’ll be much easier now to test romcal. Thanks for finding the time required to create the release!

However, I still want to use the built version because of #234. I want to update the `create_propers_from_cals.js`, which still would be a POC. Currently, I think it’d be better to _run_ romcal in order to get the required data, instead of parsing the source code.

**Comment by emagnier** [comment]
> However, I still want to use the built version because of #234. I want to update the create_propers_from_cals.js, which still would be a POC. Currently, I think it’d be better to run romcal in order to get the required data, instead of parsing the source code.

In this case, instead of installing it from NPM, you can refer to romcal NPM modules already installed locally on your computer (i.e. originally installed with a `git clone`). Do not modify your JS script, but only your `package.json` in your project, so instead of referring a specific NPM versions of the package, you can provide a path where romcal is already installed on your machine:

```json
"dependencies": {
    "@romcal/calendar.slovakia": "file:../romcal/dist/bundles/slovakia/",
    "romcal": "file:../romcal/"
  }
```

Because you manually modified your `package.json`, the first time it's probably better to delete your `node_modules/` and `package-lock.json` (potential cache issues). And then do a `npm install`.

Of course, do not forget to run `npm run build` on your `/romcal/` module directory, every time you modify something in the romcal codebase 😉

**Reply by tukusejssirs** [reply]
> > However, I still want to use the built version because of #234. I want to update the create_propers_from_cals.js, which still would be a POC. Currently, I think it’d be better to run romcal in order to get the required data, instead of parsing the source code.
>
> In this case, instead of installing it from NPM, you can refer to romcal NPM modules already installed locally on your computer (i.e. originally installed with a `git clone`). Do not modify your JS script, but only your `package.json` in your project, so instead of referring a specific NPM versions of the package, you can provide a path where romcal is already installed on your machine:

Well, I don’t think you have read #234 or you haven’t read it properly.

I need to modify `create_propers_from_cals.js` (you can find it at the end of the description of #234), because (1) it is based on an old version of romcal; (2) it directly parses the source code.

In that issue, I suggest to add a commit hook, but maybe it should be rather run when a new version is released. Nevertheless, I still think it should be using the files built locally instead of a NPM package.

> Because you manually modified your `package.json`, the first time it's probably better to delete your `node_modules/` and `package-lock.json` (potential cache issues). And then do a `npm install`.

I didn’t modify `package.json` of romcal. I invite you to re-read my previous comment, as well as #234.

**Reply by emagnier** [reply]
We are deviating from the original subject (and don't want to read again #234 which is off-topic here).

I previously gave you an example of using romcal from NPM packages (and using lazy computed metadata).

I recognize I've badly quoted a part of your previous message (and I'm sorry for this confusion). But my last message was just another example to help you, in case you are also modifying calendars or locales data from a local romcal codebase (i.e. from your computer), and want to see the built romcal data (from your local romcal package) directly in your side project (hence the suggestion of local paths in the `package.json` for your custom project).

I also forget to mention that there is an `/examples/` directory in the main `romcal` codebase that could also help.

Then, do whatever you want 🤗
